libs/partners/google-genai/langchain_google_genai/chat_models.py

```diff
@@ -0,0 +1,556 @@
+            f"Feedback: {response.prompt_feedback}"
+        )
+        generations = [generation_t(message=ai_msg_t(content=""), generation_info={})]
+    return ChatResult(generations=generations, llm_output=llm_output)
+
+
+class ChatGoogleGenerativeAI(BaseChatModel):
+    """`Google Generative AI` Chat models API.
+
+    To use you must have the google.generativeai Python package installed and
+    either:
+
+        1. The ``GOOGLE_API_KEY``` environment variable set with your API key, or
+        2. Pass your API key using the google_api_key kwarg to the ChatGoogle
+           constructor.
+
+    Example:
+        .. code-block:: python
+
+            from langchain_google_genai import ChatGoogleGenerativeAI
+            chat = ChatGoogleGenerativeAI(model="gemini-pro")
+            chat.invoke("Write me a ballad about LangChain")
+
+    """
+
+    model: str = Field(
+        ...,
+        description="""The name of the model to use.
+Supported examples:
+    - gemini-pro""",
+    )
+    max_output_tokens: int = Field(default=None, description="Max output tokens")
+
+    client: Any  #: :meta private:
+    google_api_key: Optional[str] = None
+    temperature: Optional[float] = None
+    """Run inference with this temperature. Must by in the closed
+       interval [0.0, 1.0]."""
+    top_k: Optional[int] = None
+    """Decode using top-k sampling: consider the set of top_k most probable tokens.
+       Must be positive."""
+    n: int = 1
+    """Number of chat completions to generate for each prompt. Note that the API may
+       not return the full n completions if duplicates are generated."""
+
+    _generative_model: Any  #: :meta private:
+
+    @property
+    def lc_secrets(self) -> Dict[str, str]:
+        return {"google_api_key": "GOOGLE_API_KEY"}
+
+    @property
+    def _llm_type(self) -> str:
+        return "chat-google-generative-ai"
+
+    @property
+    def _is_geminiai(self) -> bool:
+        return self.model is not None and "gemini" in self.model
+
+    @classmethod
+    def is_lc_serializable(self) -> bool:
+        return True
+
+    @root_validator()
+    def validate_environment(cls, values: Dict) -> Dict:
+        google_api_key = get_from_dict_or_env(
+            values, "google_api_key", "GOOGLE_API_KEY"
+        )
+        try:
+            import google.generativeai as genai
+
+            genai.configure(api_key=google_api_key)
+        except ImportError:
+            raise ChatGoogleGenerativeAIError(
+                "Could not import google.generativeai python package. "
+                "Please install it with `pip install google-generativeai`"
+            )
+
+        values["client"] = genai
+        if (
+            values.get("temperature") is not None
+            and not 0 <= values["temperature"] <= 1
+        ):
+            raise ValueError("temperature must be in the range [0.0, 1.0]")
+
+        if values.get("top_p") is not None and not 0 <= values["top_p"] <= 1:
```

Comment:
i think top_p isn't specified as class attribute, should it be? also candidate_count

Reply:
oh see we've got `n` for candidate count

Reply:
Actually top_p is no longer supported, my fault